pkg/resource/replication_group/delta_util.go

```diff
@@ -100,3 +113,76 @@ func unmarshalLastRequestedLDCs(desired *resource) []*svcapitypes.LogDeliveryCon
 
 	return lastRequestedConfigs
 }
+
+// multiAZRequiresUpdate returns true if the latest multi AZ status does not yet match the
+// desired state, and false otherwise
+func multiAZRequiresUpdate(desired *resource, latest *resource) bool {
+	// no preference for multi AZ specified; no update required
+	if desired.ko.Spec.MultiAZEnabled == nil {
+		return false
+	}
+
+	// API should return a non-nil value, but if it doesn't then attempt to update
+	if latest.ko.Status.MultiAZ == nil {
+		return true
+	}
+
+	// true maps to "enabled"; false maps to "disabled"
+	// this accounts for values such as "enabling" and "disabling"
+	if *desired.ko.Spec.MultiAZEnabled {
+		return *latest.ko.Status.MultiAZ != string(svcapitypes.MultiAZStatus_enabled)
+	} else {
+		return *latest.ko.Status.MultiAZ != string(svcapitypes.MultiAZStatus_disabled)
+	}
```

Comment:
Two things:

1) There should not be a `Status.MultiAZ` field. There should only be a single field: `Spec.MultiAZEnabled`

2) This kind of custom code should go in your [`updateSpecFields`](https://github.com/aws-controllers-k8s/elasticache-controller/blob/d3a3ba8e3a8da76b4b16a37b09e5675111e7f3c9/pkg/resource/replication_group/post_set_output.go#L28) function in `custom_set_output.go` and look something like this:

```go
	// MultiAZEnabled is a boolean in the Spec field but a string enum in the
	// output shape field. Here, we reconcile that difference...
	if (respRG.MultiAZ == nil) {
		// Or set this to a default value if nil isn't default?
		resource.ko.Spec.MultiAZEnabled = nil
	} else {
		resource.ko.Spec.MultiAZEnabled = (*respRG.MultiAZ == string(svcapitypes.MultiAZStatus_enabled))
	}
```

If you do the above, you don't have to worry about delta comparisons at all.

Reply:
> There should not be a Status.MultiAZ field. There should only be a single field: Spec.MultiAZEnabled

Agreed in principle, but the existence of a separate `MultiAZ` field is an ElastiCache API quirk rather than a conscious decision on our part.

> This kind of custom code should go in your updateSpecFields function in custom_set_output.go and look something like this:

I'm glad you brought this up because it's helping me externalize why we're moving away from `updateSpecFields` in many cases. 

Initially, I wanted to use `updateSpecFields` to update every single spec field during `ReadOne`, so that the generated delta code could handle everything and there would be no need for manual comparison between `desired.Spec` and `latest.Status`. 

However, `Spec` fields often don't allow us to store all the information we're getting from the API response. For example, `Spec.AutomaticFailoverEnabled` is a bool, whose associated field `Status.AutomaticFailover` is a string/enum which could be any of "enabled", "disabled", "enabling", or "disabling". We would have to map each of these four states back to a bool (square peg, round hole sort of thing), when it would make more sense to just make a decision based on the full version of the information (`Status`) since we have it anyway.

Granted, the `MultiAZ` case is a little different since at the moment, the only two possible values are "enabled" and "disabled". However, it's entirely possible that in the future ElastiCache decides to have this field update asynchronously or make any other change which could introduce additional states/values for `Status.MultiAZ`. So the suggested approach could break given these changes, and even if no service-side changes occur, I would say it's still non-ideal to force a string/enum into a bool when we can just make a decision based on the full, unadulterated information from the API response.

As far as the role of `updateSpecFields`, I think it absolutely should exist, but the function name and intended purpose should change. The majority of this function is already [calling the DescribeCacheClusters API](https://github.com/aws-controllers-k8s/elasticache-controller/blob/main/pkg/resource/replication_group/post_set_output.go#L41-L47) and populating the associated fields in the replication group object, since the latest state for some `Spec` fields cannot be retrieved directly from the `DescribeReplicationGroups API`. I would say that this should be the primary purpose of this function in the future.

So to sum up, the function currently named `updateSpecFields` is an extension of `SdkFind`/`ReadOne` in that we are still retrieving the latest state of the replication group (but via a different API). There should be limited mapping back to `Spec` fields to avoid losing information; we only do it here because this state will be lost if we don't save it back to the `latest` object. On the other hand, `modifyDelta` should look at the most complete representation of the latest state of each `Spec` field available to it (which is often in the `Status`) and make a decision based on that. Therefore I think we should keep things as-is.